From 1542d0e7c46c3aed5fa4134add96ff97c04d9774 Mon Sep 17 00:00:00 2001
From: Jan Engelhardt <jengelh@inai.de>
Date: Fri, 7 Aug 2020 21:53:22 +0000
Subject: [PATCH 07/31] build: support XSI strerror_r

UnixUserPlugin.cpp: In member function 'void UnixUserPlugin::errnoCheck(const string&, int) const':
UnixUserPlugin.cpp:920:77: error: no match for 'operator+' (operand types are 'std::__cxx11::basic_string<char>' and 'int')
  920 |   throw std::runtime_error("unable to query for user " + user + ". Error: " + retbuf);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~
      |                                                               |               |
      |                                                               |               int
      |                                                               std::__cxx11::basic_string<char>

References: KF-3318
---
 provider/plugins/UnixUserPlugin.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/provider/plugins/UnixUserPlugin.cpp b/provider/plugins/UnixUserPlugin.cpp
index 59270d988..75309d2da 100644
--- a/provider/plugins/UnixUserPlugin.cpp
+++ b/provider/plugins/UnixUserPlugin.cpp
@@ -900,7 +900,13 @@ void UnixUserPlugin::errnoCheck(const std::string &user, int e) const
 	if (e == 0)
 		return;
 	char buffer[256];
+#if defined(__GLIBC__) && (!defined(_POSIX_C_SOURCE) || \
+    _POSIX_C_SOURCE < 200112L || defined(_GNU_SOURCE))
 	auto retbuf = strerror_r(e, buffer, sizeof(buffer));
+#else
+	strerror_r(e, buffer, sizeof(buffer));
+	const char *retbuf = buffer;
+#endif
 
 	// from the getpwnam() man page: (notice the last or...)
 	//  ERRORS
-- 
2.29.2

